keps/sig-storage/596-csi-inline-volumes/README.md

```diff
@@ -290,215 +278,76 @@ Since ephemeral volume requests will participate in only the mount/unmount volum
-- [ ] Other
-  - Describe the mechanism:
-  - Will enabling / disabling the feature require downtime of the control
-    plane?
-  - Will enabling / disabling the feature require downtime or reprovisioning
-    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: CSIInlineVolume
+  - Components depending on the feature gate: kubelet
 
 ###### Does enabling the feature change any default behavior?
 
-<!--
-Any change of default behavior may be surprising to users or break existing
-automations, so be extremely careful here.
--->
+  Enabling the feature gate will **not** change the default behavior.
+  Each CSI driver must opt-in to support ephemeral inline volumes.
 
 ###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
 
-<!--
-Describe the consequences on existing workloads (e.g., if this is a runtime
-feature, can it break the existing applications?).
-
-NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
--->
+  Yes. Disabling the feature gate will disable support for inline CSI volumes.
+  It can also be disabled on a specific CSI driver by removing the "Ephemeral"
+  VolumeLifecycleMode from `CSIDriver.VolumeLifecycleModes`.
 
 ###### What happens if we reenable the feature if it was previously rolled back?
 
+  Reenabling the feature will again allow CSI drivers that opt-in to be used
+  as ephemeral inline volumes in the pod spec.
+
 ###### Are there any tests for feature enablement/disablement?
 
-<!--
-The e2e framework does not currently support enabling or disabling feature
-gates. However, unit tests in each component dealing with managing data, created
-with and without the feature, are necessary. At the very least, think about
-conversion tests if API types are being modified.
--->
+  Yes, the unit tests will test with the feature gate enabled and disabled.
 
 ### Rollout, Upgrade and Rollback Planning
 
-<!--
-This section must be completed when targeting beta to a release.
--->
-
 ###### How can a rollout or rollback fail? Can it impact already running workloads?
 
-<!--
-Try to be as paranoid as possible - e.g., what if some components will restart
-mid-rollout?
+  A rollout should not impact running workloads, since the CSI driver
+  also needs to support ephemeral inline volumes, and the pod spec would need
+  to be updated to make use of the feature.
 
-Be sure to consider highly-available clusters, where, for example,
-feature flags will be enabled on some API servers and not others during the
-rollout. Similarly, consider large clusters and how enablement/disablement
-will rollout across nodes.
--->
+  A rollback (or disabling the feature flag) may impact running workloads,
+  as any pod specs taking advantage of ephemeral inline volumes would fail
+  to start when the feature is disabled or the CSIDriver does not support it.
```

Comment:
In addition to the above, will kubelet be able to destroy a Pod / unmount its volumes when the pod uses in-line CSI volume and the feature is disabled?

IMO it will not unmount it, but I may be wrong. We should recommend deleting all Pods with in-line CSI volumes before disabling the feature.

Reply:
I don't see anything that would prevent unmounting based on feature flags. That seems like it would cause problems if kubelet restarted and had no way to unmount them.
Mounting inline CSI volumes will fail in `SetUpAt` if the feature flag is disabled:
https://github.com/kubernetes/kubernetes/blob/66445662ad7b65ffdb144b9665608f7b6a4b57a6/pkg/volume/csi/csi_mounter.go#L140-L142
But there is no such check in `TearDownAt`:
https://github.com/kubernetes/kubernetes/blob/66445662ad7b65ffdb144b9665608f7b6a4b57a6/pkg/volume/csi/csi_mounter.go#L380-L382
I'll add a note that we recommend deleting pods using this feature before disabling it though.